File: integration-tests/c/enable-defer/src/peer.c
/*
 * integration-tests/c/enable-defer -- peer.
 *
 * Simple counterpart to configurer.c: creates a normal, fully-enabled
 * DataReader on ConfigTopic immediately, then spends ~3 seconds
 * (comfortably inside the configurer's own ~4s pre-enable delay) proving it
 * observes ZERO matching -- direct evidence the deferred SEDP announcement
 * genuinely never went out while the configurer's writer was disabled.
 * After that window, waits normally for matching and 5 samples. See
 * docs/design/integration-test-tier.md for the full scenario spec.
 *
 * Required stdout markers: "Create topic: ConfigTopic", "Create reader for
 * topic: ConfigTopic", "Peer: no premature match during Ns window.", "Peer:
 * no premature match; matched and received cleanly after enable()." Any
 * failure path prints a line starting "FAIL:" and exits nonzero.
 */
#include "config_event.h"
#include "zzdds_c.h"
#include "zzdds.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define SAMPLE_TARGET 5
#define PREMATURE_CHECK_WINDOW_SEC 3
#define MATCH_TIMEOUT_MS 30000
#define RECEIVE_TIMEOUT_MS 20000
#define POLL_PERIOD_MS 20

static uint32_t parse_domain(int argc, char **argv) {
    for (int i = 1; i < argc - 1; i++) {
        if (strcmp(argv[i], "-d") == 0 || strcmp(argv[i], "--domain") == 0) {
            return (uint32_t)strtoul(argv[i + 1], NULL, 10);
        }
    }
    return 0;
}

int main(int argc, char **argv) {
    uint32_t domain_id = parse_domain(argc, argv);

    zzdds_DomainParticipantFactory factory = zzdds_create_factory();
    if (zzdds_factory_is_nil(factory)) {
        fprintf(stderr, "FAIL: createFactory() failed\n");
        return 1;
    }
    DDS_DomainParticipantFactory dds_factory = zzdds_DomainParticipantFactory_as_DDS_DomainParticipantFactory(factory);

    DDS_DomainParticipant dp = DDS_DomainParticipantFactory_create_participant(dds_factory, domain_id, NULL, NULL, 0);
    if (!dp) {
        fprintf(stderr, "FAIL: create_participant() failed on domain %u\n", domain_id);
        return 1;
    }

    if (ConfigEventTypeSupport_register(dp, "ConfigEvent") != DDS_RETCODE_OK) {
        fprintf(stderr, "FAIL: register ConfigEventTypeSupport failed\n");
        return 1;
    }

    DDS_Topic topic = DDS_DomainParticipant_create_topic(dp, "ConfigTopic", "ConfigEvent", NULL, NULL, 0);
    if (!topic) {
        fprintf(stderr, "FAIL: create_topic(ConfigTopic) failed\n");
        return 1;
    }
    printf("Create topic: ConfigTopic\n");

    DDS_Subscriber sub = DDS_DomainParticipant_create_subscriber(dp, NULL, NULL, 0);
    if (!sub) {
        fprintf(stderr, "FAIL: create_subscriber() failed\n");
        return 1;
    }

    DDS_DataReaderQos dr_qos;
    DDS_Subscriber_get_default_datareader_qos(sub, &dr_qos);
    dr_qos.reliability.kind = DDS_ReliabilityQosPolicyKind_RELIABLE_RELIABILITY_QOS;
    dr_qos.history.kind = DDS_HistoryQosPolicyKind_KEEP_ALL_HISTORY_QOS;

    DDS_TopicDescription desc = zzdds_topic_as_description(topic);
    DDS_DataReader dr = DDS_Subscriber_create_datareader(sub, desc, &dr_qos, NULL, 0);
    if (!dr) {
        fprintf(stderr, "FAIL: create_datareader(ConfigTopic) failed\n");
        return 1;
    }
    printf("Create reader for topic: ConfigTopic\n");

    ConfigEventDataReader reader;
    ConfigEventDataReader_init(&reader, dr);

    /* Core assertion: for a window comfortably inside the configurer's own
     * pre-enable delay, matched-current-count must stay exactly 0 -- direct
     * proof the deferred SEDP announcement genuinely never went out while
     * the configurer's writer was disabled. */
    DDS_SubscriptionMatchedStatus status;
    for (int waited_ms = 0; waited_ms < PREMATURE_CHECK_WINDOW_SEC * 1000; waited_ms += POLL_PERIOD_MS) {
        if (DDS_DataReader_get_subscription_matched_status(dr, &status) != DDS_RETCODE_OK) {
            fprintf(stderr, "FAIL: get_subscription_matched_status() failed\n");
            return 1;
        }
        if (status.current_count != 0) {
            fprintf(stderr, "FAIL: matched before enable() was called -- deferred SEDP announcement isn't working (current_count=%d)\n", status.current_count);
            return 1;
        }
        usleep(POLL_PERIOD_MS * 1000);
    }
    printf("Peer: no premature match during %ds window.\n", PREMATURE_CHECK_WINDOW_SEC);

    /* Now wait normally for the real match, once the configurer enables. */
    bool matched = false;
    for (int waited_ms = 0; !matched; waited_ms += POLL_PERIOD_MS) {
        if (waited_ms >= MATCH_TIMEOUT_MS) {
            fprintf(stderr, "FAIL: never matched within %ds of the premature-match window ending\n", MATCH_TIMEOUT_MS / 1000);
            return 1;
        }
        if (DDS_DataReader_get_subscription_matched_status(dr, &status) != DDS_RETCODE_OK) {
            fprintf(stderr, "FAIL: get_subscription_matched_status() failed\n");
            return 1;
        }
        if (status.current_count > 0) {
            matched = true;
        } else {
            usleep(POLL_PERIOD_MS * 1000);
        }
    }

    int received = 0;
    int32_t last_seq = -1;
    ConfigEvent values[SAMPLE_TARGET];
    DDS_SampleInfo infos[SAMPLE_TARGET];
    for (int waited_ms = 0; received < SAMPLE_TARGET; waited_ms += POLL_PERIOD_MS) {
        if (waited_ms >= RECEIVE_TIMEOUT_MS) {
            fprintf(stderr, "FAIL: did not receive all %d samples within %ds (got %d)\n", SAMPLE_TARGET, RECEIVE_TIMEOUT_MS / 1000, received);
            return 1;
        }
        int n = ConfigEventDataReader_take_n(&reader, values, infos, SAMPLE_TARGET,
                                              DDS_ANY_SAMPLE_STATE, DDS_ANY_VIEW_STATE, DDS_ANY_INSTANCE_STATE);
        for (int i = 0; i < n; i++) {
            if (!infos[i].valid_data) continue;
            if (values[i].seq != last_seq + 1) {
                fprintf(stderr, "FAIL: out-of-order sample, expected seq=%d got seq=%d\n", last_seq + 1, values[i].seq);
                return 1;
            }
            last_seq = values[i].seq;
            received++;
        }
        if (received < SAMPLE_TARGET) usleep(POLL_PERIOD_MS * 1000);
    }

    printf("Peer: no premature match; matched and received cleanly after enable().\n");

    DDS_DomainParticipantFactory_delete_participant(dds_factory, dp);
    zzdds_destroy_factory(factory);
    return 0;
}

File: integration-tests/zig/ignore-entities/peer.zig
//! zig/ignore-entities -- peer. Simple counterpart to ignorer.zig: creates a
//! writer for ControlTopic and TopicIgnoredTopic, and a reader for
//! SubscriptionIgnoredTopic, plus a writer that gets matched by the
//! ignorer's throwaway "probe" reader on PublicationIgnoredTopic. See
//! docs/design/integration-test-tier.md for the full scenario spec.
//!
//! IMPORTANT: this file deliberately does NOT assert that its own
//! match-count fields ever drop to zero. ignore_topic()/ignore_publication()
//! are a strictly one-sided, local filter on the ignorer's participant (see
//! ignorer.zig's header comment) -- from this process's own SEDP
//! perspective, its TopicIgnoredTopic/PublicationIgnoredTopic writers
//! legitimately stay matched to the ignorer's reader the entire time,
//! exactly as if nothing had been ignored at all. The one thing this
//! process CAN and does verify from its own side is
//! SubscriptionIgnoredTopic: ignore_subscription() is called on the
//! *writer's* participant (ignorer), so it's ignorer's own writer that
//! never adds this process's reader as a matched proxy -- meaning this
//! reader, despite itself reporting a normal SEDP match, must never
//! actually receive any of the real writer's samples.
//!
//! Required stdout markers: "Peer: ready.", "Peer: SubscriptionIgnoredTopic
//! reader received zero samples from the real (post-ignore) writer.",
//! "Peer: done." Any failure path prints a line starting "FAIL:" and exits
//! nonzero.

const std = @import("std");
const zzdds = @import("zzdds");
const DDS = @import("zzdds_generated").DDS;
const ignore_event_gen = @import("ignore_event_gen");

const SAMPLE_COUNT: i32 = 5;
// NOT long enough to guarantee the ignorer has finished every ignore_*()
// call -- its own PROBE_MATCH_TIMEOUT_MS (45s) applies twice, sequentially,
// for the publication and subscription probes, so its true worst case is
// well over a minute. Raised from 6s to match this file's own
// MATCH_TIMEOUT_NS convention as a meaningfully better (not watertight)
// margin: the SubscriptionIgnoredTopic check below can still report a false
// pass -- zero samples because the real (post-ignore) writer hasn't been
// created yet, not because ignore_subscription() worked -- if the ignorer
// is still deep in its own probe/settle choreography when this window
// closes. A fully watertight fix needs an explicit cross-process signal
// (e.g. a dedicated marker topic) rather than a fixed sleep; not done here
// to avoid adding a worst-case 100+s wait on top of an already
// CI-budget-constrained suite (found via Greptile review; see
// docs/roadmap.md's discovery-latency entry for the same underlying "how
// long is long enough" tension).
const SETTLE_WINDOW_NS: u64 = 20 * std.time.ns_per_s;
const MATCH_TIMEOUT_NS: i64 = 20 * std.time.ns_per_s;
const DRAIN_TIMEOUT_NS: i64 = 15 * std.time.ns_per_s;
const POLL_PERIOD_NS: u64 = 20 * std.time.ns_per_ms;

fn monoNs(io: std.Io) i64 {
    return @intCast(std.Io.Clock.awake.now(io).nanoseconds);
}

fn sleepNs(io: std.Io, ns: u64) void {
    (std.Io.Clock.Duration{ .raw = .{ .nanoseconds = @intCast(ns) }, .clock = .awake }).sleep(io) catch {};
}

fn parseDomain(process_args: std.process.Args) u32 {
    var it = std.process.Args.Iterator.init(process_args);
    _ = it.skip();
    while (it.next()) |arg| {
        if (std.mem.eql(u8, arg, "-d") or std.mem.eql(u8, arg, "--domain")) {
            const v = it.next() orelse continue;
            return std.fmt.parseInt(u32, v, 10) catch 0;
        }
    }
    return 0;
}

pub fn main(init: std.process.Init) !void {
    const io = init.io;
    var gpa = std.heap.DebugAllocator(.{}){};
    defer _ = gpa.deinit();
    const alloc = gpa.allocator();

    const domain_id = parseDomain(init.minimal.args);

    var factory = zzdds.createFactory() catch {
        std.debug.print("FAIL: createFactory() failed\n", .{});
        std.process.exit(1);
    };
    defer factory.deinit();
    const dpf = factory.toDDSFactory();

    const dp = dpf.create_participant(domain_id, .{}, null, 0);
    if (dp.ptr == zzdds.dcps.NIL_PTR) {
        std.debug.print("FAIL: create_participant() failed on domain {d}\n", .{domain_id});
        std.process.exit(1);
    }
    defer _ = dpf.delete_participant(dp);

    var ts_alloc = alloc;
    if (!zzdds.registerTypeSupport(dp, "IgnoreEvent", .{
        .ctx = @ptrCast(&ts_alloc),
        .compute_key_hash = ignore_event_gen.IgnoreEvent.computeKeyHashFromCdr,
        .compute_key_hash_key_only = ignore_event_gen.IgnoreEvent.computeKeyHashFromCdrKeyOnly,
    })) {
        std.debug.print("FAIL: registerTypeSupport() failed\n", .{});
        std.process.exit(1);
    }

    const control_topic = dp.create_topic("ControlTopic", "IgnoreEvent", .{}, null, 0);
    const topic_ignored_topic = dp.create_topic("TopicIgnoredTopic", "IgnoreEvent", .{}, null, 0);
    const pub_ignored_topic = dp.create_topic("PublicationIgnoredTopic", "IgnoreEvent", .{}, null, 0);
    const sub_ignored_topic = dp.create_topic("SubscriptionIgnoredTopic", "IgnoreEvent", .{}, null, 0);
    if (control_topic.ptr == zzdds.dcps.NIL_PTR or topic_ignored_topic.ptr == zzdds.dcps.NIL_PTR or
        pub_ignored_topic.ptr == zzdds.dcps.NIL_PTR or sub_ignored_topic.ptr == zzdds.dcps.NIL_PTR)
    {
        std.debug.print("FAIL: create_topic() failed\n", .{});
        std.process.exit(1);
    }

    const publisher = dp.create_publisher(.{}, null, 0);
    const subscriber = dp.create_subscriber(.{}, null, 0);
    if (publisher.ptr == zzdds.dcps.NIL_PTR or subscriber.ptr == zzdds.dcps.NIL_PTR) {
        std.debug.print("FAIL: create_publisher()/create_subscriber() failed\n", .{});
        std.process.exit(1);
    }

    var dw_qos = DDS.DataWriterQos{};
    dw_qos.reliability.kind = .RELIABLE_RELIABILITY_QOS;
    dw_qos.history.kind = .KEEP_ALL_HISTORY_QOS;
    var dr_qos = DDS.DataReaderQos{};
    dr_qos.reliability.kind = .RELIABLE_RELIABILITY_QOS;
    dr_qos.history.kind = .KEEP_ALL_HISTORY_QOS;

    const control_dw = publisher.create_datawriter(control_topic, dw_qos, null, 0);
    const topic_ignored_dw = publisher.create_datawriter(topic_ignored_topic, dw_qos, null, 0);
    const pub_ignored_dw = publisher.create_datawriter(pub_ignored_topic, dw_qos, null, 0);
    const sub_ignored_dr = subscriber.create_datareader(sub_ignored_topic.vtable.as_TopicDescription(sub_ignored_topic.ptr), dr_qos, null, 0);
    if (control_dw.ptr == zzdds.dcps.NIL_PTR or topic_ignored_dw.ptr == zzdds.dcps.NIL_PTR or
        pub_ignored_dw.ptr == zzdds.dcps.NIL_PTR or sub_ignored_dr.ptr == zzdds.dcps.NIL_PTR)
    {
        std.debug.print("FAIL: create_datawriter()/create_datareader() failed\n", .{});
        std.process.exit(1);
    }
    std.debug.print("Peer: ready.\n", .{});

    // Write a few samples on TopicIgnoredTopic and PublicationIgnoredTopic
    // right away -- both writers exist continuously from here on, giving
    // the ignorer's readers every real opportunity to (wrongly) match.
    const topic_ignored_writer = ignore_event_gen.IgnoreEventDataWriter.init(topic_ignored_dw, alloc);
    const pub_ignored_writer = ignore_event_gen.IgnoreEventDataWriter.init(pub_ignored_dw, alloc);
    var seq: i32 = 0;
    while (seq < SAMPLE_COUNT) : (seq += 1) {
        topic_ignored_writer.write(.{ .seq = seq }, 0) catch {};
        pub_ignored_writer.write(.{ .seq = seq }, 0) catch {};
    }

    // Let the ignorer's full choreography (participant-discover-and-ignore,
    // publication probe-then-ignore, subscription probe-then-ignore) run to
    // completion. No assertion on topic_ignored_dw's or pub_ignored_dw's own
    // match-count here -- see this file's header comment for why that
    // would be asserting something ignore_topic()/ignore_publication()
    // never promised.
    sleepNs(io, SETTLE_WINDOW_NS);

    // The one thing this process's own side CAN verify: ignore_subscription()
    // was called on ignorer's *writer* participant, so its real
    // (post-ignore) writer never adds this reader as a matched proxy --
    // meaning this reader, however its own SEDP match status reports
    // itself, must never actually receive any of that writer's samples.
    const sub_ignored_reader = ignore_event_gen.IgnoreEventDataReader.init(sub_ignored_dr, alloc);
    var taken: std.ArrayListUnmanaged(ignore_event_gen.IgnoreEventDataReader.SampledValue) = .empty;
    defer taken.deinit(alloc);
    _ = try sub_ignored_reader.take(&taken, -1, DDS.ANY_SAMPLE_STATE, DDS.ANY_VIEW_STATE, DDS.ANY_INSTANCE_STATE);
    if (taken.items.len != 0) {
        std.debug.print("FAIL: SubscriptionIgnoredTopic reader received {d} samples, expected 0\n", .{taken.items.len});
        std.process.exit(1);
    }
    std.debug.print("Peer: SubscriptionIgnoredTopic reader received zero samples from the real (post-ignore) writer.\n", .{});

    // Normal ControlTopic round-trip, matching every other scenario's
    // sanity-check convention.
    var control_status: DDS.PublicationMatchedStatus = undefined;
    const match_deadline = monoNs(io) + MATCH_TIMEOUT_NS;
    var control_matched = false;
    while (!control_matched) {
        if (monoNs(io) > match_deadline) {
            std.debug.print("FAIL: ControlTopic writer never matched within {d}s\n", .{@divTrunc(MATCH_TIMEOUT_NS, std.time.ns_per_s)});
            std.process.exit(1);
        }
        if (control_dw.get_publication_matched_status(&control_status) != DDS.RETCODE_OK) {
            std.debug.print("FAIL: get_publication_matched_status(control) failed\n", .{});
            std.process.exit(1);
        }
        if (control_status.current_count > 0) control_matched = true else sleepNs(io, POLL_PERIOD_NS);
    }
    const control_writer = ignore_event_gen.IgnoreEventDataWriter.init(control_dw, alloc);
    seq = 0;
    while (seq < SAMPLE_COUNT) : (seq += 1) {
        control_writer.write(.{ .seq = seq }, 0) catch {
            std.debug.print("FAIL: write(ControlTopic) failed at seq={d}\n", .{seq});
            std.process.exit(1);
        };
    }

    // Standard teardown-safety: wait for the ignorer to disconnect before
    // deleting, matching every other scenario's precedent.
    const drain_deadline = monoNs(io) + DRAIN_TIMEOUT_NS;
    while (true) {
        if (control_dw.get_publication_matched_status(&control_status) != DDS.RETCODE_OK) {
            std.debug.print("FAIL: get_publication_matched_status(control) failed\n", .{});
            std.process.exit(1);
        }
        if (control_status.current_count == 0) break;
        if (monoNs(io) > drain_deadline) {
            std.debug.print("FAIL: ignorer did not disconnect within {d}s\n", .{@divTrunc(DRAIN_TIMEOUT_NS, std.time.ns_per_s)});
            std.process.exit(1);
        }
        sleepNs(io, POLL_PERIOD_NS);
    }

    std.debug.print("Peer: done.\n", .{});
}
